worker/upload.go

```diff
@@ -0,0 +1,836 @@
+	for _, q := range u.pool {
+		close(q.stopChan)
+	}
+	close(u.statsStopChan)
+}
+
+func (u *uploader) blockHeight() uint64 {
+	u.poolMu.Lock()
+	defer u.poolMu.Unlock()
+	return u.bh
+}
+
+func (u *uploader) newUpload() uploadID {
+	var id uploadID
+	frand.Read(id[:])
+	return id
+}
+
+func (u *uploader) finishUpload(id uploadID) {
+	u.poolMu.Lock()
+	defer u.poolMu.Unlock()
+	for _, q := range u.pool {
+		q.finish(id)
+	}
+}
+
+func (u *uploader) upload(ctx context.Context, r io.Reader, contracts []api.ContractMetadata, rs api.RedundancySettings, bh uint64) (o object.Object, used map[types.PublicKey]types.FileContractID, err error) {
+	// add tracing
+	ctx, span := tracing.Tracer.Start(ctx, "uploader.upload")
+	defer func() {
+		if err != nil {
+			span.RecordError(err)
+		}
+		span.End()
+	}()
+
+	// sanity check redundancy
+	if rs.TotalShards > len(contracts) {
+		err := errors.New("not enough contracts to meet redundancy")
+		span.RecordError(err)
+		return object.Object{}, nil, err
+	}
+
+	// refresh contracts
+	u.updatePool(contracts, bh)
+
+	// keep track of used hosts
+	used = make(map[types.PublicKey]types.FileContractID)
+
+	// build contracts map
+	h2c := make(map[types.PublicKey]types.FileContractID)
+	for _, c := range contracts {
+		h2c[c.HostKey] = c.ID
+	}
+
+	// initialize the encryption key
+	o.Key = object.GenerateEncryptionKey()
+	cr := o.Key.Encrypt(r)
+
+	for {
+		proceed, err := func() (bool, error) {
+			// ensure the upload ctx is cancelled
+			ctx, cancel := context.WithCancel(ctx)
+			defer cancel()
+
+			// create span
+			ctx, span := tracing.Tracer.Start(ctx, "uploader.slab")
+			defer span.End()
+
+			// initialize upload
+			id := u.newUpload()
+			defer u.finishUpload(id)
+			span.SetAttributes(attribute.Stringer("id", id))
+
+			// create slab
+			slab := object.Slab{
+				Key:       object.GenerateEncryptionKey(),
+				MinShards: uint8(rs.MinShards),
+			}
+
+			// read slab data
+			buf := make([]byte, slab.Length())
+			length, err := io.ReadFull(io.LimitReader(cr, int64(slab.Length())), buf)
+			if err == io.EOF {
+				return false, nil
+			} else if err != nil && err != io.ErrUnexpectedEOF {
+				return false, err
+			}
+			span.AddEvent("shards read")
+
+			// encode and encrypt the shards
+			shards := make([][]byte, rs.TotalShards)
+			slab.Encode(buf, shards)
+			span.AddEvent("shards encoded")
+			slab.Encrypt(shards)
+			span.AddEvent("shards encrypted")
+
+			// upload the shards
+			slab.Shards, err = u.uploadSectors(ctx, id, shards)
+			if err != nil {
+				return false, err
+			}
+
+			// add the slab
+			o.Slabs = append(o.Slabs, object.SlabSlice{
+				Slab:   slab,
+				Offset: 0,
+				Length: uint32(length),
+			})
+
+			// update used hosts
+			for _, ss := range slab.Shards {
+				if _, ok := used[ss.Host]; !ok {
+					used[ss.Host] = h2c[ss.Host]
+				}
+			}
+			return true, nil
+		}()
+		if err != nil {
+			return object.Object{}, nil, err
+		}
+		if !proceed {
+			break
+		}
```

Comment:
Potential f/u idea. A host that we uploaded a shard to could theoretically move on to the next slab. But only the next one and not the one after that. Also only hosts that have successfully been used already. Because we want to reserve all unused hosts to potentially use them for overdriving the last shard.
That could reduce the idle time we see between slabs.

Reply:
That's a good idea, I'm not sure I fully understand why we want to reserve all unused hosts to potentially overdrive the last shard? Your idea sounds like a "slab-buffer" where we process at most two slabs at a time to me, more or less. Essentially separating the `io.ReadFull` and encoding/encrypting from the sector upload, which I think is a good idea because we'll smoothen out the uploads that way.